File: iac/aws/observability/inspec/controls/thanos.rb
# Copyright (C) 2020 Nicolas Lamirault <nicolas.lamirault@gmail.com>

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

key_id = attribute("thanos_key_id")
bucket_name = attribute("thanos_bucket_name")
role_name = attribute("thanos_role_name")
policy_name = attribute("thanos_policy_name")


control "thanos-1" do
  impact 1.0

  title "Check that S3 bucket exist"

  tag platform: "AWS"
  tag category: 'Service'
  tag resource: "thanos"
  tag effort: 0.2

  describe aws_s3_bucket(bucket_name) do
    it { should exist }
    it { should_not be_public }
    its('tags') { should include(
      'made-by' => 'terraform',
      'project' => 'foo',
      'service' => 'thanos'
    )}
    it { should have_default_encryption_enabled }
  end

end

control "thanos-2" do
  impact 1.0

  title "Check that Kms key exist and tags are correctly set"

  tag platform: "AWS"
  tag category: 'Service'
  tag resource: "thanos"
  tag effort: 0.2

  describe aws_kms_key(key_id) do
    it { should exist }
    # its('tags') { should include(
    #   'made-by' => 'terraform',
    #   'project' => 'portefaix',
    #   'service' => 'thanos'
    # )}
  end

end

control "thanos-3" do
  impact 1.0

  title "Check IAM"

  tag platform: "AWS"
  tag category: 'Service'
  tag resource: "thanos"
  tag effort: 0.2

  describe aws_iam_role(role_name: role_name) do
    it { should exist }
    # its('tags') { should include(
    #   'made-by' => 'terraform',
    #   'project' => 'portefaix',
    #   'service' => 'thanos'
    # )}
  end

  describe aws_iam_policy(policy_name: policy_name) do
    it { should exist }
  end

end

File: diagrams/thanos.py
#!/usr/bin/python3

# Copyright (C) 2021 Nicolas Lamirault <nicolas.lamirault@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse

import diagrams
from diagrams.k8s import compute
from diagrams.k8s import network
from diagrams.k8s import others
from diagrams.k8s import podconfig
from diagrams.k8s import rbac
from diagrams.k8s import storage

import cloud


def cloud_provider_resources(cloud_provider):
    iam = cloud.iam(cloud_provider)
    bucket = cloud.bucket(cloud_provider)
    disk = cloud.disk(cloud_provider)
    return iam, bucket, disk


def architecture(cloud_provider, output, direction):
    with diagrams.Diagram("thanos_%s" % cloud_provider, direction="TB", show=False):
        with diagrams.Cluster("Cloud Platform"):

            iam, bucket, disk = cloud_provider_resources(cloud_provider)

            sc = storage.StorageClass()
            compact_pvc = storage.PVC("compact")
            compact_pv = storage.PV("compact")
            sc << compact_pvc
            compact_pvc >> compact_pv

            store_pvc = storage.PVC("store")
            store_pv = storage.PV("store")
            sc << store_pvc
            store_pvc >> store_pv

            sc >> [store_pv, compact_pv]
            disk << [store_pv, compact_pv]

            with diagrams.Cluster("Kubernetes Cluster"):
                clusterRole = rbac.ClusterRole()
                clusterRoleBinding = rbac.ClusterRoleBinding()

                with diagrams.Cluster("monitoring"):
                    sa = rbac.ServiceAccount()
                    clusterRole << clusterRoleBinding >> sa
                    role = rbac.Role()
                    roleBinding = rbac.RoleBinding()
                    role << roleBinding >> sa
                    # secret = podconfig.Secret("storage")

                    compact_svc = network.Service("compact")
                    compact_sts = compute.StatefulSet("compact")
                    compact_pod = compute.Pod("compact")
                    compact_sm = others.CRD("servicemonitor")

                    compact_sts >> compact_pod >> iam  # secret
                    compact_svc >> compact_pod
                    compact_pod >> compact_pv
                    compact_sm >> compact_svc

                    query_svc = network.Service("query")
                    query_deploy = compute.Deployment("query")
                    query_pod = compute.Pod("query")
                    query_sm = others.CRD("servicemonitor")

                    store_svc = network.Service("store")
                    store_sts = compute.StatefulSet("store")
                    store_pod = compute.Pod("store")
                    store_sm = others.CRD("servicemonitor")

                    query_deploy >> query_pod
                    query_svc >> query_pod
                    query_sm >> query_svc
                    query_pod >> store_svc

                    [store_sts, store_svc] >> store_pod
                    store_pod >> iam  # secret
                    store_pod >> store_pv
                    store_sm >> store_svc

                    sa >> iam
                    compact_pod >> bucket
                    store_pod >> bucket


def main(cloud_provider, output, direction):
    architecture(args.cloud, args.output, args.direction)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-o",
        "--output",
        type=str,
        choices=__outformats,
        default="png",
        help="Output format",
    )
    parser.add_argument(
        "-d",
        "--direction",
        type=str,
        choices=__directions,
        default="LR",
        help="Diagram direction",
    )
    parser.add_argument(
        "-c",
        "--cloud",
        type=str,
        choices=__cloud_providers,
        # default="gcp",
        help="Cloud provider",
    )
    args = parser.parse_args()
    main(args.cloud, args.output, args.direction)
